Code/Framework/AzCore/AzCore/std/typetraits/has_member_function.h

```diff
@@ -5,76 +5,141 @@
  * SPDX-License-Identifier: Apache-2.0 OR MIT
  *
  */
-#ifndef AZSTD_TYPE_TRAITS_HAS_MEMBER_FUNCTION_INCLUDED
-#define AZSTD_TYPE_TRAITS_HAS_MEMBER_FUNCTION_INCLUDED
+#pragma once
 
-#include <AzCore/std/typetraits/is_class.h>
+#include <AzCore/std/function/invoke.h>
+#include <AzCore/std/typetraits/conditional.h>
+#include <AzCore/std/typetraits/remove_cvref.h>
+#include <AzCore/variadic.h>
 
 /**
 * Helper to create checkers for member function inside a class.
+* Explanation of how checks are done:
+* 1. Direct Member check:
+* This validates that the class itself has the member and that is not being detected from any base class
+* The check is done by using the box_member template to enforce that the member function pointer
+* is actually from the supplied class type and not from a base type
 *
+* 2. Member Invocable with Parameters and returns Result check:
+* This validates that the member exist on the class and if it is a non-static member
+* it can be invoked with a signature of (<class name instance>, Params...)
+* or if it is a static member it can be invoked with a signature of (Params...)
+* and returns the result type R
+*/
+// Used to append options to the AZStd::is_invocable_r_v below
+#define AZ_PREFIX_ARG_WITH_COMMA(_Arg) , _Arg
+
+
+namespace AZStd::HasMemberInternal
+{
+    // Function declaration only used in unevaluated context
+    // in order to enforce that the member function pointer
+    // is a direct member of the class
+    template<class T, class MemType, MemType Member, class... Args>
+    auto direct_member_box() -> AZStd::invoke_result_t<MemType, T, Args...>;
+}
+
+/**
+* AZ_HAS_MEMBER detects if the member exists directly on the class and
+* can be invoked with the specified function name and signature
+*
+* NOTE: This does not include base classes in the check
 * Ex. AZ_HAS_MEMBER(IsReadyMember,IsReady,void,())
 *     struct A {};
 *     struct B
 *     {
 *        void IsReady();
 *     };
+*     struct Der : B {};
 *     HasIsReadyMember<A>::value (== false)
 *     HasIsReadyMember<B>::value (== true)
+*     HasIsReadyMember<Der>::value (== false)
```

Comment:
The name in the example doesn't match the one defined above (`IsReadyMember` vs. `HasIsReadyMember`).

Reply:
That is because the macro always prepends the word "Has" to the first argument of AZ_HAS_MEMBER macro.